torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +381,761 @@ bool IsBlockReturnTypeSame(Node* n) {
+      } else {
+        if (static_required) {
+          return c10::nullopt;
+        }
+        symbol_vector.emplace_back(c10::ShapeSymbol::newSymbol());
+      }
+    }
+    return c10::SymbolicShape(symbol_vector);
+  }
+  return c10::nullopt;
+}
+
+void ProcessConstantValueMap(Node* n, int opset_version) {
+  // Update ConstantValueMap on node outputs from onnx shape inference
+  // For outputs, only update static shapes. For input, we update symbolic
+  // shapes also. ONNX If can have different types on different branches, skip
+  // here.
+  for (auto i = 0; i < n->outputs().size(); i++) {
+    if (TensorTypePtr output_type = n->output(i)->type()->cast<TensorType>()) {
+      if (output_type->dim().has_value()) {
+        size_t rank = static_cast<size_t>(output_type->dim().value());
+        ConstantValueMap::SetRank(n->output(i)->debugName(), rank);
+        auto output_type_size = output_type->sizes();
+        auto shape = GetSymbolicShapeFromVaryingTypes(
+            output_type_size, /* static_required = */ true);
+        if (shape.has_value()) {
+          UpdateShape(n->output(i), shape.value());
+        }
+      }
+    }
+  }
+  // Update ConstantValueMap on node inputs from onnx shape inference.
+  // ListConstruct input is handled here, no need to have a per-op based
+  // process.
+  for (auto i = 0; i < n->inputs().size(); i++) {
+    if (TensorTypePtr input_type = n->input(i)->type()->cast<TensorType>()) {
+      if (input_type->dim().has_value()) {
+        size_t rank = static_cast<size_t>(input_type->dim().value());
+        ConstantValueMap::SetRank(n->input(i)->debugName(), rank);
+        auto input_type_size = input_type->sizes();
+        auto shape = GetSymbolicShapeFromVaryingTypes(input_type_size);
+        if (!ConstantValueMap::HasShape(n->input(i)->debugName())) {
+          if (shape.has_value()) {
+            UpdateShape(n->input(i), shape.value());
+          }
+        }
+      }
+    } else if (n->input(i)->type()->cast<ListType>()) {
+      auto lc_node = n->input(i)->node();
+      // TODO: check ListConstruct array vs. tensor?
+      auto rank = lc_node->inputs().size();
+      auto shape_size = GetValueFromListConstructNode(lc_node);
+      if (shape_size.has_value()) {
+        auto shape_vector = shape_size.value();
+        auto options = c10::TensorOptions().dtype(at::kLong).device(at::kCPU);
+        auto shape_value_size = static_cast<int64_t>(shape_vector.size());
+        auto f =
+            at::from_blob(shape_vector.data(), {shape_value_size}, at::kLong)
+                .to(at::kCPU);
+        // Need copy here
+        at::Tensor f_copy = at::empty({shape_value_size}, options);
+        f_copy.copy_(f);
+        ConstantValueMap::SetValue(n->input(i)->debugName(), f_copy);
+        UpdateShapeFromVector(n->input(i), {shape_value_size});
+      } else {
+        UpdateShapeFromVector(n->input(i), {static_cast<int64_t>(rank)});
```

Comment:
wonder if this is always true, for case where inputs are tensors

Reply:
I change the logic here: we only consider ListConstruct IntType case. For tensorType, we don't update value/rank.